Drop -dbcache limit

Sjors opened the pull request:
Due to recent UTXO set growth, the current maximum value for `-dbcache` of 16GB is ~just months away from being~ insufficient (for those who wish to complete IBD with the UTXO set held in RAM).

This drops the limit. It also adds a warning that it's up to users to check that they have enough RAM.

Fixes #28249.

---

A previous version of this PR increased the maximum to 64GB. It also made startup abort if the value provided is too high, rather than quietly round it down. But this didn't get much support.

jamesob commented:
> This can easily lead to OOM

Wonder if it's worth warning/failing if the dbcache param exceeds available memory.

Sjors commented:
My recollection is that an IBD without a single flush is faster, but have not benched this in recent years.

We can always _recommend_ a lower setting if that turns out to be better. By not increasing it to infinity at least we constrain the worst case here, e.g. if it turns out the failures you mean get exponentially worse as the UTXO set keeps growing. While at the same time not removing the existing ability to sync in one go.

Sjors commented on doc/release-notes-25358.md:
```diff
@@ -0,0 +1,5 @@
+Node
+------
+
+- The maximum value for `-dbcache` has been increased to 32,000 MiB due to recent UTXO set growth. Be sure to check that you have enough RAM before increasing this.
```
That was not the intended implication. If they feel like increasing it, they should check their RAM first.

Sjors replied:
I dropped the "check RAM" sentence.

andrewtoth commented:
> My recollection is that an IBD without a single flush is faster, but have not benched this in recent years.

A single flush throughout the entire multi-hour IBD will not make a noticeable difference to a user. My initial version of https://github.com/bitcoin/bitcoin/pull/28280 had 3 extra pointers per cache entry, which caused the `dbcache` limit to be hit. It was still >2% faster than master without a single flush.

TheCharlatan commented:
Concept ACK.

henrikhk commented:
Concept ACK.

BenWestgate commented:
Concept ACK.

Except for:

> make startup abort if the value provided is too high, rather than quietly rounding it down.

This breaks scripts that set dbcache to a % of available RAM.

And is inconsistent with it rounding up silently rather than fail to start if the value is too low.

0xB10C commented:
Concept ACK on increasing the maximum `dbcache`. 

I had to manually bump the dbcache to run `bitcoin-cli verifychain 4 0` last year (https://github.com/bitcoin/bitcoin/pull/24851#issuecomment-1214040286). This came close to using 32 GB and might not be enough anymore (or soon - I haven't checked again). Maybe worth considering going directly to 64 GB here to be able to actually use the `verifychain 4 0` RPC?

darosior commented:
Concept ACK on giving more options.

Sjors commented:
> This breaks scripts that set dbcache to a % of available RAM.

Do you know any example of people doing that?

In general I think we should fail to startup if the user provides an invalid configuration option. I believe this has been gradually introduced for various other options too.

At the same time I don't want to needless break scripts out there.

From my own experience I was very surprised to find that this rounding down behavior happened, which is which I prefer an error. But we could also print a warning in the debug log.

> Maybe worth considering going directly to 64 GB here to be able to actually use the `verifychain 4 0` RPC?

@sipa any thoughts on a sane maximum?

BenWestgate commented:
> > This breaks scripts that set dbcache to a % of available RAM.
> 
> Do you know any example of people doing that?

My [project](https://github.com/BenWestgate/Bails/blob/master/share/applications/bitcoin-qt.desktop) Bails puts this in `bitcoin-qt.desktop` as well as `.config/autostart/Bitcoin/bitcoin.desktop:`
`-dbcache=$(($(grep Available /proc/meminfo | sed s/[^0-9]//g)/1024-2000))`

It's an alpha with about 20 users, makes a portable USB stick node that keeps Bitcoin Core up to date, none probably own a 64GB computer, but it will break if they do.

I found a couple other projects with the idea:
https://github.com/Start9Labs/bitcoind-wrapper/issues/39
https://github.com/JWWeatherman/yeticold/issues/145
https://github.com/epiccurious/bitcoin-core-node-builder/issues/38

They'll waste time if they miss the change notes.
Maybe a one time warning would be better when the value rounds up or down? It took me a long time to notice that it did that.

jonatack reviewed (commented):
Concept ACK

> In general I think we should fail to startup if the user provides an invalid configuration option, as this often leads to problems. I believe this has been gradually introduced for various other options too (see also #16545).

https://github.com/bitcoin/bitcoin/pull/27632 was another recent example of switching from a warning to raising an error. It's too easy for users, even experienced ones, not to notice a warning line in the log.

Sjors commented on doc/release-notes-25358.md:
```diff
@@ -0,0 +1,5 @@
+Node
+------
+
+- The maximum value for `-dbcache` has been increased to 32,000 MiB due to recent UTXO set growth. Be sure to check that you have enough RAM before increasing this.
```
I might also add something like "Even with sufficient RAM the maximum setting may not be optimal."

Sjors replied:
Taken the suggested text diff.

Sjors commented on doc/release-notes-25358.md:
```diff
@@ -0,0 +1,5 @@
+Node
```
Release notes get edited later anyway, but I'll change once I retouch.

(waiting for more inputs in the discussion on warning/failure and the max size)

Sjors replied:
Updated.

BenWestgate commented:
Perhaps it should be set so high that you can't easily buy a Desktop with more RAM than the limit.

I could go to Microcenter today and walk out with a 64GB gaming tower, or a Xeon workstation with 128GB, likely a 64GB laptop as well.

Lots of guides on internet (foolishly/arbitrarily) suggest setting it to half the total RAM.

Sjors commented on src/init.cpp:
```diff
@@ -925,6 +925,10 @@ bool AppInitParameterInteraction(const ArgsManager& args)
     }
 
+    if (args.GetIntArg("-dbcache", nDefaultDbCache) >  nMaxDbCache) {
+        return InitError(Untranslated(strprintf("-dbcache must be at most %d MiB", nMaxDbCache)));
```
That indeed seems like the best approach. Silently rounding down is annoying, but doesn't break anything, so a warning should be fine.

Sjors replied:
I'm leaving this error as is for now, unless there are other examples of breaking software.

I would also rather drop `nMaxDbCache` entirely than quietly enforce it.

luke-jr commented:
Why do we have an upper limit at all?

>My understanding was that we can't measure available RAM reliably.

#19873 does exactly that

luke-jr commented on src/txdb.h:
```diff
@@ -38,7 +38,7 @@ static const int64_t nDefaultDbCache = 450;
 static const int64_t nDefaultDbBatchSize = 16 << 20;
 //! max. -dbcache (MiB)
-static const int64_t nMaxDbCache = sizeof(void*) > 4 ? 16384 : 1024;
+static const int64_t nMaxDbCache = sizeof(void*) > 4 ? 32768 : 1024;
```
Might as well make it `constexpr` while you're touching it

Sjors replied:
Done

Sjors commented on src/init.cpp:
```diff
@@ -925,6 +925,10 @@ bool AppInitParameterInteraction(const ArgsManager& args)
 
+    if (args.GetIntArg("-dbcache", nDefaultDbCache) >  nMaxDbCache) {
+        return InitError(Untranslated(strprintf("-dbcache must be at most %d MiB", nMaxDbCache)));
+    }
```
Added a test.

Sjors commented on src/init.cpp:
```diff
@@ -446,7 +446,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-datadir=<dir>", "Specify data directory", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-dbbatchsize", strprintf("Maximum database write batch size in bytes (default: %u)", nDefaultDbBatchSize), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-dbcache=<n>", strprintf("Maximum database cache size <n> MiB (%d to %d, default: %d). In addition, unused mempool memory is shared for this cache (see -maxmempool).", nMinDbCache, nMaxDbCache, nDefaultDbCache), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-dbcache=<n>", strprintf("Maximum database cache size <n> MiB (%d to %d, default: %d). Make sure you have enough RAM. In addition, unused mempool memory is shared for this cache (see -maxmempool).", nMinDbCache, nMaxDbCache, nDefaultDbCache), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
Taken

Sjors commented:
Rebased, incorporated some of the text feedback above. Dropped the RAM warning from the release note, since that's already in the help text.

> Why do we have an upper limit at all?

@sipa seemed worried about users going overboard with this setting: https://github.com/bitcoin/bitcoin/issues/28249#issuecomment-1695572615

tdb3 reviewed (commented):
crACK

Sjors commented:
Fixed broken `feature_config_args.py` test, taking into account the lower limit on 32 bit systems.

Sjors commented:
I was still waiting for someone to suggest a new value. But it seems fine to bump to 64 GB in order to support `verifychain 4 0`, so added that.

@maflcko do you prefer to drop the limit? We should not quietly change a user setting, so imo an invalid value should explicitly fail, rather than silently fail.

> Also, CI still fails

Indeed. I copied the `is_64bits = sys.maxsize > 2**32` check from `netutil.py`, but it seems the check doesn't actually work. So there might be a bug in `netutil.py`.

Meanwhile the `multiprocess, i686` CI compiler seems to think it's 32 bit, so the `izeof(void*)` check in `txdb.h` seems broken too. (which could be another reason to just drop the limit entirely)

Sjors commented:
We could expose whether the program is compiled for 32 or 64 bit in the test framework... but I just added a regex for the error message instead.

cbergqvist reviewed (approved):
ACK 39a3d48.

### Concept
Adding this kind of hard error might make it more annoying to run old Bitcoin Core software in the far future. (But the consequence of a few extra flushes should be fine).

### Tested
Starting `bitcoind` with -dbcache set to low value, max value, and beyond max value with expected results.

Ran `./test/functional/feature_config_args.py` with and without modified expected assert message, with expected results.

Sjors commented on src/node/caches.cpp:
```diff
@@ -13,7 +13,7 @@ CacheSizes CalculateCacheSizes(const ArgsManager& args, size_t n_indexes)
     int64_t nTotalCache = (args.GetIntArg("-dbcache", nDefaultDbCache) << 20);
     nTotalCache = std::max(nTotalCache, nMinDbCache << 20); // total cache cannot be less than nMinDbCache
-    nTotalCache = std::min(nTotalCache, nMaxDbCache << 20); // total cache cannot be greater than nMaxDbcache
+    assert(nTotalCache <= nMaxDbCache << 20);
```
I'll add a comment if I have to retouch.

Sjors replied:
No longer needed since the limit is dropped now.

Sjors commented:
It seems there's more enthusiasm for dropping this limit entirely, so I did that. This also impacts the GUI (but none of the translated strings), cc @hebasto.

For future reference, c016405dadcdc645f31dc77cb0a92e2e10fe358a + 39a3d48a4ae87fd659a58ce93109a722fc436dd8 is the last version of this PR which _does_ have a limit.

Sjors commented on src/qt/optionsdialog.cpp:
```diff
@@ -94,8 +94,7 @@ OptionsDialog::OptionsDialog(QWidget* parent, bool enableWallet)
     /* Main elements init */
-    ui->databaseCache->setMinimum(nMinDbCache);
-    ui->databaseCache->setMaximum(nMaxDbCache);
+    ui->databaseCache->setRange(nMinDbCache, std::numeric_limits<int>::max());
```
Even on a 32 bit system the maximum would be 2 terabyte, so not a practical concern imo.

Sjors commented:
@sipa since 32 bit systems can't have more than 4 GiB of RAM, the "Make sure you have enough RAM" wording in the documentation should cover that. Though perhaps "Stay at least 1 GiB below your RAM size" may be safer, also because of swap behaviour (see e.g. https://github.com/bitcoin/bitcoin/issues/29603#issuecomment-1991010152)

With something like #19873 we could warn the use if they pick a value too close to the limit (on both 32 and 64 bit systems).

>  wildly too large

We could suggest an optimum, but without further experimentation we don't know where that is.

Ideally we would measure performance in real time and flush when it starts degrading (unless the user really insists that we don't).

cbergqvist commented:
Built 5f313495eea36b566109894697b4a90085c53113 and ran bitcoin-qt because I was worried that setting a very high max value might cause issues with the QT UI control. It's not some kind of horizontal slider that goes crazy, but rather a textbox with up/down arrows. It ignores key inputs that would make the value go further than `std::numeric_limits<int>::max()`. So all good from that perspective.

![bitcoin-qt_screenshot](https://github.com/bitcoin/bitcoin/assets/581308/61038c42-9e7f-47b9-9e99-551237f9c772)

hebasto commented:
> This also impacts the GUI (but none of the translated strings), cc @hebasto.

5f313495eea36b566109894697b4a90085c53113: changes to the GUI code look good to me. It might worth considering adding the "Make sure you have enough RAM." warning to the GUI as well (as a follow-up).

Sjors commented:
> "Make sure you have enough RAM." warning to the GUI as well (as a follow-up).

Will do if I need to retouch.

andrewtoth commented:
After #28233 keeping the entire utxo set in memory will also help connect new blocks as fast as possible, not just for IBD.

Sjors commented:
Rebased to make it easier to test the interaction with (merged) #28233. However, any potential benefit goes away after the first restart. The only time -dbcache is filled beyond its previous 16GB limit is during IDB.

In order to benefit from having the full UTXO set in RAM after a restart, we'd have to proactive load the entire UTXO set into -dbcache if space permits.  

Added "Make sure you have enough RAM." warning to the GUI.

andrewtoth commented:
> The only time -dbcache is filled beyond its previous 16GB limit is during IDB

I'm sure it will happen if you restart your node now and leave it running for long enough ;)

> In order to benefit from having the full UTXO set in RAM after a restart, we'd have to proactively load the entire UTXO set into -dbcache if space permits.

I did make a crude attempt at that years ago https://github.com/bitcoin/bitcoin/pull/18941. Not sure it is worth it.

BenWestgate reviewed (approved):
crACK

sdaftuar commented on doc/release-notes-25358.md:
```diff
@@ -0,0 +1,6 @@
+Updated settings
```
Just noticed the filename here should be fixed (25358 vs 28358).

sdaftuar reviewed (commented):
Concept ACK.

I don't see much of a philosophical difference between having no limit and having a limit of 16GB on systems where a user has less than 16GB actually available...  If we're already requiring users to do something smart, we can keep doing that -- and if we are worried about someone using an inappropriate value we can try to solve that separately (maybe by testing that we can allocate that much at startup or something?).

BenWestgate commented:
> If we are worried about someone using an inappropriate value we can try to solve that separately (maybe by testing that we can allocate that much at startup or something?).

The best would be never setting it greater than total RAM. `-dbcache=<total RAM>` is still higher than optimal as there will be applications in memory that won't be swapped to disk including Bitcoin. But measuring Available RAM or swap is precarious as it changes.

Since other reviewers mentioned it's only wildly excessive values are seriously detrimental with swap, capping to max RAM is sensible.

> In order to benefit from having the full UTXO set in RAM after a restart, we'd have to proactively load the entire UTXO set into -dbcache if space permits. Or maybe fill half the available space with coins, picked randomly.

Some have used `cat .bitcoin/chainstate/* > /dev/null` to make a node that's far behind start syncing faster. Because of the disk io to read the chainstate folder, there's a trade off between initial speed and medium-term speed. A node mere hours behind shouldn't read the entire UTXO set into memory. But one that's weeks behind will reach tip faster if it does.

> Though perhaps "Stay at least 1 GiB below your RAM size" may be safer, also because of swap behaviour (see e.g. https://github.com/bitcoin/bitcoin/issues/29603#issuecomment-1991010152)

That's barely safe without swap, Bitcoin-QT and the OS will use most of the 1 GiB offset leaving it pretty close to locking up.

Without a swap file, an offset from total ram (1-2 GiB?) may avoid OOM, but this becomes arbitrary and may limit performance if set too low. It's better users w/o swap who set it too high OOM, than to limit performance of systems w/ swap that may benefit from more.

So total memory is still the best limit, if any is enforced.

> We could suggest an optimum, but without further experimentation we don't know where that is.

I've done a lot of tests on 4 GB and 8 GB RAM machines with zram swap and the -datadir on USB sticks and `available_RAM - 2 GiB` is about optimal for keeping those systems responsive. (With `-dbcache=<available_RAM - 2 GiB>` both the 4 GB and 8 GB can unlock the screen and manipulate the desktop with a 5-10 seconds lag, had the user locked screen and left Bitcoin to sync several hours. With 4 GB RAM, this is approximately our default `-dbcache=450`.) If -dbcache is too high with a slow swap the machine may take minutes to respond. That's true even when the value is set lower than total or available memory. But whether that lag is undesirable or not depends on whether the user intends to use the machine or not before sync completes.

Sjors committed:
validation: drop maximum -dbcache

Fixes #28249

Sjors commented:
@BenWestgate thanks for the analysis. If we want to add a recommendation to the help text, I suggest we do that in another PR though. It's probably going to take multiple people doing benchmark on wildly different machines.

Checking against RAM can also be done in a followup.

I rebased and changed the .md file name.

jlopp commented:
Concept ACK.

For reference, I ran several sync tests with different dbcache sizes to see how much of an effect they have on performance. A good 25% improvement vs default when my machine abstained from flushing to disk. https://blog.lopp.net/effects-dbcache-size-bitcoin-node-sync-speed/

Sjors commented:
@jlopp thanks for the testing and your article.

I assume "to reach a given block height with a 28 GB dbcache size" you did _not_ run with this patch? I'd be curious to see the result.

And if you feel like battle testing this further, you could make a custom signet, produce a similar UTXO set to mainnet (maybe #24952 can help), increase it to a few hundred gigabyte and then see if and where there's a value `-dbcache` where performance peaks and goes down.

tdb3 reviewed (approved):
ACK bb3b980dfd96d9a89e6b04aef2859ce0555c6807

Seems reasonable that a user explicitly setting `dbcache` would have the freedom (and responsibility) to allocate as much memory as desirable.

Rebased this PR branch on top of master (d661e2b1b771abafb0b152842d775d3150032230).
```
2024-09-06T03:37:15Z Cache configuration:
2024-09-06T03:37:15Z * Using 2.0 MiB for block index database
2024-09-06T03:37:15Z * Using 8.0 MiB for chain state database
2024-09-06T03:37:15Z * Using 42998.0 MiB for in-memory UTXO set (plus up to 286.1 MiB of unused mempool space)
```

## Sync-from-Genesis Tests

Ran three sync tests with another local node (running 27.1, stuck at 860036) on a LAN, with both nodes configured to only `connect` to each other.  Tests ran with an inexpensive DRAMless SATA SSD.

Observed a significantly faster sync with dbcache > 16384, with the tradeoff of having to spend more time writing out to disk (e.g. on exit).  Overall, seems like an advantage to give the user the choice.

### Default `dbcache`

Time to reach block 860036: 27 hours 29 minutes
Time to exit bitcoind (writing out): 4 minutes

### `dbcache=16384`

Time to reach block 860036: 8 hours 31 minutes
Time to exit bitcoind (writing out): 8 minutes (cache was 4285.7MiB at the time)
Significant write out seen at blocks 804273 and 849619.

### `dbcache=43008`

Time to reach block 860036: 6 hours 42 minutes
Time to exit bitcoind (writing out): 110 minutes (cache was 26691.3MiB at the time)
Significant write out seen only when exiting bitcoins.

Value of 43008 (42GiB) was chosen as a balance of being above expected cache high water mark during sync and low enough to not impact OS RAM needs for other programs.

BenWestgate commented:
> ### `dbcache=43008`
> Time to reach block 860036: 6 hours 42 minutes Time to exit bitcoind (writing out): 110 minutes (cache was 26691.3MiB at the time) Significant write out seen only when exiting bitcoins.

The long writing out would go significantly faster on a higher performance SSD, it is also CPU-intensive as it's compressing the 27GB in memory. So the faster initial sync is a win.

andrewtoth commented:
If we merge https://github.com/bitcoin/bitcoin/pull/30611, then write outs will be more evenly distributed throughout the sync and will not all happen on shutdown with a high dbcache.

l0rinc commented:
Concept ACK, I'll run some tests later

l0rinc commented:
I ran benchmarks to evaluate the impact of removing the limit on IBD performance.
The tests were conducted on an Intel Core i7-7700 CPU, 64 GB of RAM, and HDD storage.
I synced up 3 times to block height 600,000 using various `-dbcache` settings: 2 GB, 5 GB, 10 GB, 20 GB, 30 GB, and 40 GB.

<details>
<summary>benchmark</summary>

```bash
hyperfine \
--runs 3 \
--export-json /mnt/ibd_dbcache.json \
--parameter-list DBCACHE 2048,5120,10240,20480,30720,40960 \
--prepare 'rm -rf /mnt/BitcoinData/*' \
'COMMIT={COMMIT} ./build/src/bitcoind -datadir=/mnt/BitcoinData -stopatheight=600000 -dbcache={DBCACHE} -printtoconsole=0'
```

</details>

### Results:
* 2 GB: 4 hours, 39 minutes, 52 seconds
* 5 GB: 4 hours, 5 minutes, 50 seconds
* 10 GB: 3 hours, 51 minutes, 35 seconds
* 20 GB: 3 hours, 48 minutes, 25 seconds
* 30 GB: 3 hours, 45 minutes, 53 seconds
* 40 GB: 3 hours, 45 minutes, 43 seconds

<img width="651" alt="image" src="https://github.com/user-attachments/assets/686de93a-f2a7-43d8-beb6-5e818b3a4f7f">

<details>
<summary>Details</summary>

dbcache | seconds
--------------------
2048 | 18565.20009
2048 | 16450.20701
2048 | 15362.17072
5120 | 14805.76276
5120 | 14797.24122
5120 | 14646.52724
10240 | 14453.19471
10240 | 13883.38449
10240 | 13347.80065
20480 | 14022.99233
20480 | 13658.22833
20480 | 13433.18961
30720 | 13688.31978
30720 | 13595.47899
30720 | 13376.19299
40960 | 13692.83482
40960 | 13672.43475
40960 | 13263.82877

</details>

### Findings:

* Increasing `-dbcache` from 2 GB to 10 GB resulted in a significant improvement in IBD time, reducing it by about 48 minutes (~17% reduction).
* Beyond 10 GB, the performance gains diminished, with less than a 2% improvement when increasing from 10 GB to 40 GB (approximately 6 minutes saved).

Based on this data I think removing the limit may not be substantiated, since setting `-dbcache` beyond 10 GB offers minimal additional benefits (for IBD at least, maybe the difference would be bigger with pruned nodes).

BenWestgate commented:
> I ran benchmarks to evaluate the impact of removing the limit on IBD performance. ... I synced up 3 times to block height 600,000...

Thanks for this benchmark. The current block height is 861,090 and the size of the utxo set has more than doubled since 600,000 so this is why you do not see performance improvements from dbcache > 10 GB.

> Based on this data I think removing the limit may not be substantiated...

Try syncing to 860,000 there should be performance benefit up to 30 GiB dbcache as some syncs grew to use 27GiB cache.

BenWestgate commented on src/qt/forms/optionsdialog.ui:
```diff
@@ -105,7 +105,7 @@
           <widget class="QLabel" name="databaseCacheLabel">
            <property name="toolTip">
-            <string extracomment="Tooltip text for Options window setting that sets the size of the database cache. Explains the corresponding effects of increasing/decreasing this value.">Maximum database cache size. A larger cache can contribute to faster sync, after which the benefit is less pronounced for most use cases. Lowering the cache size will reduce memory usage. Unused mempool memory  ...
+            <string extracomment="Tooltip text for Options window setting that sets the size of the database cache. Explains the corresponding effects of increasing/decreasing this value.">Maximum database cache size. Make sure you have enough RAM. A larger cache can contribute to faster sync, after which the benefit is less pronounced for most use cases. Lowering the cache size will reduce memor ...
```
Nit: "Make sure you have enough available RAM." seems more correct.

BenWestgate reviewed (commented):
crACK bb3b980dfd96d9a89e6b04aef2859ce0555c6807.

Setting `dbcache` below available RAM performs better and swaps less than setting it below total RAM, which may even crash the machine or OOM kill other applications.

I feel the tool tip and argument texts should say "available RAM" but it's not a blocker for me.

Sjors commented on src/init.cpp:
```diff
@@ -474,7 +474,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-datadir=<dir>", "Specify data directory", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-dbbatchsize", strprintf("Maximum database write batch size in bytes (default: %u)", nDefaultDbBatchSize), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-dbcache=<n>", strprintf("Maximum database cache size <n> MiB (%d to %d, default: %d). In addition, unused mempool memory is shared for this cache (see -maxmempool).", nMinDbCache, nMaxDbCache, nDefaultDbCache), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-dbcache=<n>", strprintf("Maximum database cache size <n> MiB (minimum %d, default: %d). Make sure you have enough RAM. In addition, unused memory allocated to the mempool is shared with this cache (see -maxmempool).", nMinDbCache, nDefaultDbCache), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
I'm not sure what "available" means here? The amount of RAM used by other processes can vary wildly, especially on a desktop computer.

In any case, I prefer to leave changes to the text to a followup.

BenWestgate replied:
> I'm not sure what "available" means here? 

From [man free](https://man7.org/linux/man-pages/man1/free.1.html):

> **available**
> Estimation of how much memory is available for starting new applications, without swapping. Unlike the data provided by the cache or free fields, this field takes into account page cache and also that not all reclaimable memory slabs will be reclaimed due to items being in use

So having "enough RAM" will work but having "enough available memory" is a more accurate number to stay under to avoid swapping or OOM errors.

Someone may assume with 16 GB RAM they can safely set dbcache to 8-15 GB but on most desktop systems that would swap or crash.

achow101 commented:
ACK bb3b980dfd96d9a89e6b04aef2859ce0555c6807

jonatack reviewed (commented):
ACK bb3b980dfd96d9a89e6b04aef2859ce0555c6807

Agee with more freedom/responsibility for the user and the ability to reduce flushing to disk for better IBD performance.

It's probably a good idea to avoid using strike-through tags in PR descriptions, as they aren't rendered well by the merge script and look like emphasis instead.

    Due to recent UTXO set growth, the current maximum value for `-dbcache` of 16GB is ~just months away from being~ insufficient (for those who wish to complete IBD with the UTXO set held in RAM).

l0rinc commented:
Unsurprisingly (should we document this?), setting `-prune` makes the `-dbcache` value less important:

<details>
<summary>benchmark</summary>

```bash
// 6d7f24595b08b8d1eba53e648533bcf87c30b48f
hyperfine \
--runs 2 \
--export-json /mnt/ibd_dbcache_prune.json \
--parameter-list DBCACHE 10240,16384,20480,30720,40960 \
--prepare 'rm -rf /mnt/BitcoinData/*' \
'./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache={DBCACHE}'
```

</details>

```bash
Benchmark 1: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=10240
  Time (mean ± σ):     54047.663 s ± 196.467 s    [User: 48343.781 s, System: 5183.690 s]
  Range (min … max):   53908.739 s … 54186.586 s    2 runs

Benchmark 2: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=16384
  Time (mean ± σ):     54971.917 s ± 33.433 s    [User: 47761.439 s, System: 4482.604 s]
  Range (min … max):   54948.276 s … 54995.558 s    2 runs

Benchmark 3: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=20480
  Time (mean ± σ):     54625.234 s ± 372.120 s    [User: 47902.577 s, System: 4480.376 s]
  Range (min … max):   54362.105 s … 54888.362 s    2 runs

Benchmark 4: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=30720
  Time (mean ± σ):     54996.086 s ± 500.933 s    [User: 47059.146 s, System: 3619.912 s]
  Range (min … max):   54641.873 s … 55350.299 s    2 runs

Benchmark 5: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=40960
  Time (mean ± σ):     56834.434 s ± 569.458 s    [User: 47748.418 s, System: 3785.085 s]
  Range (min … max):   56431.766 s … 57237.102 s    2 runs
```

BenWestgate commented:
Didn’t #28280 resolve this?

On Fri, Sep 20, 2024 at 05:16, l0rinc ***@***.***(mailto:On Fri, Sep 20, 2024 at 05:16, l0rinc <<a href=)> wrote:

> Unsurprisingly (should we document this?), setting -prune makes the -dbcache value less important:
>
> benchmark
>
> // 6d7f24595b08b8d1eba53e648533bcf87c30b48f
> hyperfine \
> --runs 2 \
> --export-json /mnt/ibd_dbcache_prune.json \
> --parameter-list DBCACHE 10240,16384,20480,30720,40960 \
> --prepare
>
> '
>
> rm -rf /mnt/BitcoinData/*
>
> '
>
> \
>
> '
>
> ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache={DBCACHE}
>
> '
>
> Benchmark 1: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=10240
>   Time (mean ± σ):     54047.663 s ± 196.467 s    [User: 48343.781 s, System: 5183.690 s]
>   Range (min … max):   53908.739 s … 54186.586 s    2 runs
>
> Benchmark 2: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=16384
>   Time (mean ± σ):     54971.917 s ± 33.433 s    [User: 47761.439 s, System: 4482.604 s]
>   Range (min … max):   54948.276 s … 54995.558 s    2 runs
>
> Benchmark 3: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=20480
>   Time (mean ± σ):     54625.234 s ± 372.120 s    [User: 47902.577 s, System: 4480.376 s]
>   Range (min … max):   54362.105 s … 54888.362 s    2 runs
>
> Benchmark 4: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=30720
>   Time (mean ± σ):     54996.086 s ± 500.933 s    [User: 47059.146 s, System: 3619.912 s]
>   Range (min … max):   54641.873 s … 55350.299 s    2 runs
>
> Benchmark 5: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -prune=550 -dbcache=40960
>   Time (mean ± σ):     56834.434 s ± 569.458 s    [User: 47748.418 s, System: 3785.085 s]
>   Range (min … max):   56431.766 s … 57237.102 s    2 runs
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/28358#issuecomment-2363371383), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ARQZ6FZCU5O7OMP5K6UEAWDZXPYWVAVCNFSM6AAAAAA4BUB2WOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNRTGM3TCMZYGM).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

l0rinc commented:
> Didn’t #28280 resolve this?

That's why I was checking it, I've rebased it on [9cb9651d92ddb5d92724f6a52440601c7a0bbcf8](https://github.com/bitcoin/bitcoin/commit/9cb9651d92ddb5d92724f6a52440601c7a0bbcf8) before running it with `-prune`.
@andrewtoth is this what you expected to happen, or was I measuring something wrong here?

<details>
<summary>Details</summary>

```bash
# git rev-parse HEAD
6d7f24595b08b8d1eba53e648533bcf87c30b48f

# git log --grep="bitcoin/bitcoin#28280"
commit 27a770b34b8f1dbb84760f442edb3e23a0c2420b
Merge: 0f68a05c08 589db872e1
Author: Ava Chow <github@achow101.com>
Date:   Wed Aug 7 20:06:39 2024 -0400

    Merge bitcoin/bitcoin#28280: Don't empty dbcache on prune flushes: >30% faster IBD
```

</details>

----


I ran it since without `-prune ` as well since, and the differences are still minuscule (5% better than before - likely because of the HDD of the server):

<details>
<summary>benchmark</summary>

```bash
hyperfine \
--runs 1 \
--export-json /mnt/ibd_dbcache_full.json \
--parameter-list DBCACHE 10240,16384,20480,30720,40960 \
--prepare 'rm -rf /mnt/BitcoinData/*' \
'./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -dbcache={DBCACHE}'
```

</details>

```bash
Benchmark 1: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -dbcache=10240
  Time (abs ≡):        43454.890 s               [User: 36019.244 s, System: 3368.345 s]

Benchmark 2: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -dbcache=16384
  Time (abs ≡):        40668.815 s               [User: 35550.351 s, System: 2825.697 s]

Benchmark 3: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -dbcache=20480
  Time (abs ≡):        40441.252 s               [User: 35192.236 s, System: 2627.069 s]

Benchmark 4: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -dbcache=30720
  Time (abs ≡):        40075.069 s               [User: 34405.162 s, System: 1974.380 s]

Benchmark 5: ./build/src/bitcoind -datadir=/mnt/BitcoinData -printtoconsole=0 -stopatheight=860000 -dbcache=40960
  Time (abs ≡):        38508.128 s               [User: 33951.127 s, System: 1771.498 s]
```

andrewtoth commented:
I really think there's too much variance when syncing with untrusted peers.
